### PR TITLE
mat64: make Inner take Vector type

### PR DESCRIPTION
**Pull request by kortschak**
This needs tests for Inc != 1. I'm thinking try all four of `[]struct{ incX, incY int }{{1, 1}, {1,2}, {2, 1}, {2, 2}}` in a loop with the given data. This needs then a helper to construct those vectors from the []float that are in the test struct.

### CODE REVIEW

**Comment by kortschak** [comment]
@btracey

File: mat64/inner.go [hunk 1]
```diff
@@ -29,26 +29,65 @@ func Inner(x []float64, A Matrix, y []float64) float64 {
 	switch b := A.(type) {
 	case RawSymmetricer:
 		bmat := b.RawSymmetric()
```
Comment (btracey) [inline mat64/inner.go]: This should probably be "case RawSymmetricer, RawTriangular", and then get the data slice and stride as variables. Also, we should check that the Uplo is correct.

**Comment by btracey** [comment]
@kortschak , I think the functions I implemented in BLAS are nice for this. It puts weird unique values in the off-elements, which makes debugging easier when the code is incorrect.

File: mat64/inner.go (hunk 1 above)
Comment (btracey) [inline mat64/inner.go]: I guess it's more complicated with Triangular, because we support both shapes.

Comment (kortschak) [inline mat64/inner.go]: Can't do `case RawSymmetricer, RawTriangular:`. We can do separate cases and send the results to a func.

**Comment by kortschak** [comment]
I'll poison the off index elements with NaN.

File: mat64/inner.go (hunk 1 above)
Comment (btracey) [inline mat64/inner.go]: Yes, sorry, I meant that informally. Sorry for not being specific.

File: mat64/inner.go [hunk 2]
```diff
@@ -11,13 +11,13 @@ import "github.com/gonum/internal/asm"
 // between vectors x and y with matrix A. This is only a true inner product if
 // A is symmetric positive definite, though the operation works for any matrix A.
 //
-// Inner panics if len(x) != m or len(y) != n when A is an m x n matrix.
-func Inner(x []float64, A Matrix, y []float64) float64 {
+// Inner panics if x.Len != m or y.Len != n when A is an m x n matrix.
```
Comment (btracey) [inline mat64/inner.go]: Should these be "x.Len()" since it's a function and not a field?

**Comment by kortschak** [comment]
Just read your comments there, I'll do that.

File: mat64/inner.go (hunk 2 above)
Comment (kortschak) [inline mat64/inner.go]: Conventionally no.

File: mat64/inner.go (hunk 1 above)
Comment (kortschak) [inline mat64/inner.go]: I think the Triangular handling is for another PR. I'll add the Uplo check though.

**Comment by btracey** [comment]
The code in Inner looks correct to me

File: mat64/inner.go (hunk 1 above)
Comment (btracey) [inline mat64/inner.go]: SGTM

**Comment by kortschak** [comment]
OK, PTAL

**Comment by btracey** [comment]
LGTM

**Comment by btracey** [comment]
Thanks

**Pull request merged**